File: web/src/main/java/no/nav/sbl/dialogarena/modiabrukerdialog/web/rest/oppgave/OppgaveController.java
package no.nav.sbl.dialogarena.modiabrukerdialog.web.rest.oppgave;

import no.nav.nav.sbl.dialogarena.modiabrukerdialog.api.domain.Temagruppe;
import no.nav.nav.sbl.dialogarena.modiabrukerdialog.api.service.LeggTilbakeOppgaveIGsakRequest;
import no.nav.nav.sbl.dialogarena.modiabrukerdialog.api.service.OppgaveBehandlingService;
import no.nav.nav.sbl.dialogarena.modiabrukerdialog.api.utils.featuretoggling.FeatureToggle;
import no.nav.sbl.dialogarena.modiabrukerdialog.web.rest.util.CookieUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Response;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static no.nav.metrics.MetricsFactory.createEvent;
import static no.nav.nav.sbl.dialogarena.modiabrukerdialog.api.utils.featuretoggling.Feature.DELVISE_SVAR;

@Path("/oppgaver/{id}")
@Produces(APPLICATION_JSON)
public class OppgaveController {

    private static Logger logger = LoggerFactory.getLogger(OppgaveController.class);

    private final OppgaveBehandlingService oppgaveBehandlingService;

    @Inject
    public OppgaveController(OppgaveBehandlingService oppgaveBehandlingService) {
        this.oppgaveBehandlingService = oppgaveBehandlingService;
    }

    @POST
    @Path("/leggTilbake")
    @Consumes(APPLICATION_JSON)
    public Response leggTilbake(@PathParam("id") String oppgaveId, @Context HttpServletRequest httpRequest, LeggTilbakeRESTRequest request) {
        if (!FeatureToggle.visFeature(DELVISE_SVAR)) {
            return Response.serverError().status(Response.Status.NOT_IMPLEMENTED).build();
        }

        LeggTilbakeOppgaveIGsakRequest leggTilbakeOppgaveIGsakRequest = new LeggTilbakeOppgaveIGsakRequest()
                .withOppgaveId(oppgaveId)
                .withBeskrivelse(request.beskrivelse)
                .withTemagruppe(getTemagruppefraRequest(request.temagruppe))
                .withSaksbehandlersValgteEnhet(CookieUtil.getSaksbehandlersValgteEnhet(httpRequest));

        try {
            oppgaveBehandlingService.leggTilbakeOppgaveIGsak(leggTilbakeOppgaveIGsakRequest);
        } catch (RuntimeException exception) {
            throw handterRuntimeFeil(exception);
        }
        createEvent("hendelse.oppgavecontroller.leggtilbake.fullfort").report();
        return Response.ok("{\"message\": \"Success\"}").build();
    }

    private Temagruppe getTemagruppefraRequest(String temagruppe){
        try {
            return Temagruppe.valueOf(temagruppe);
        } catch(IllegalArgumentException e) {
            throw new IllegalArgumentException("Ugyldig temagruppe: " + temagruppe);
        }
    }

    private RuntimeException handterRuntimeFeil(RuntimeException exception) {
        logger.error("Feil ved legging av oppgave tilbake til GSAK", exception);
        createEvent("hendelse.oppgavecontroller.leggtilbake.runtime-exception").report();
        return exception;
    }

}

File: web/src/main/java/no/nav/sbl/dialogarena/modiabrukerdialog/web/rest/oppgave/OppgaveController.kt
package no.nav.sbl.dialogarena.modiabrukerdialog.web.rest.oppgave

import no.nav.sbl.dialogarena.modiabrukerdialog.api.domain.Oppgave
import no.nav.sbl.dialogarena.modiabrukerdialog.api.domain.Temagruppe
import no.nav.sbl.dialogarena.modiabrukerdialog.api.service.HenvendelseUtsendingService
import no.nav.sbl.dialogarena.modiabrukerdialog.api.service.LeggTilbakeOppgaveIGsakRequest
import no.nav.sbl.dialogarena.modiabrukerdialog.api.service.OppgaveBehandlingService
import no.nav.sbl.dialogarena.modiabrukerdialog.api.utils.RestUtils
import no.nav.sbl.dialogarena.modiabrukerdialog.api.utils.http.CookieUtil
import no.nav.sbl.dialogarena.modiabrukerdialog.tilgangskontroll.Policies
import no.nav.sbl.dialogarena.modiabrukerdialog.tilgangskontroll.Tilgangskontroll
import no.nav.sbl.dialogarena.modiabrukerdialog.web.service.plukkoppgave.PlukkOppgaveService
import no.nav.sbl.dialogarena.naudit.Audit
import no.nav.sbl.dialogarena.naudit.Audit.Action.READ
import no.nav.sbl.dialogarena.naudit.Audit.Action.UPDATE
import no.nav.sbl.dialogarena.naudit.AuditIdentifier
import no.nav.sbl.dialogarena.naudit.AuditResources.Person.Henvendelse
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest

private val logger = LoggerFactory.getLogger(OppgaveController::class.java)
private const val AARSAK_PREFIX = "Oppgave lagt tilbake. Årsak: "

@RestController
@RequestMapping("/rest/oppgaver")
class OppgaveController @Autowired constructor(
        private val oppgaveBehandlingService: OppgaveBehandlingService,
        private val plukkOppgaveService: PlukkOppgaveService,
        private val henvendelseUtsendingService: HenvendelseUtsendingService,
        private val tilgangkontroll: Tilgangskontroll
) {

    @PostMapping("/legg-tilbake")
    fun leggTilbake(httpRequest: HttpServletRequest, @RequestBody request: LeggTilbakeRequest): ResponseEntity<Void> {
        return tilgangkontroll
                .check(Policies.tilgangTilModia)
                .get(Audit.describe(UPDATE, Henvendelse.Oppgave.LeggTilbake, AuditIdentifier.OPPGAVE_ID to request.oppgaveId)) {
                    val valgtEnhet = RestUtils.hentValgtEnhet(httpRequest)
                    val leggTilbakeOppgaveIGsakRequest = lagLeggTilbakeRequest(request, valgtEnhet)

                    try {
                        oppgaveBehandlingService.leggTilbakeOppgaveIGsak(leggTilbakeOppgaveIGsakRequest)
                        if (request.type == LeggTilbakeAarsak.FeilTema) {
                            henvendelseUtsendingService.oppdaterTemagruppe(request.traadId, request.temagruppe.toString())
                        }
                    } catch (exception: RuntimeException) {
                        throw handterRuntimeFeil(exception)
                    }

                    ResponseEntity(HttpStatus.OK)
                }
    }

    @PostMapping("/plukk/{temagruppe}")
    fun plukkOppgaver(@PathVariable("temagruppe") temagruppe: String, httpRequest: HttpServletRequest): List<OppgaveDTO> {
        return tilgangkontroll
                .check(Policies.tilgangTilModia)
                .check(Policies.kanPlukkeOppgave)
                .get(Audit.describe(READ, Henvendelse.Oppgave.Plukk, AuditIdentifier.TEMAGRUPPE to temagruppe)) {
                    val tildelteOppgaver = oppgaveBehandlingService.finnTildelteOppgaverIGsak()
                    if (tildelteOppgaver.isNotEmpty()) {
                        tildelteOppgaver
                    } else {
                        plukkOppgaveService
                                .plukkOppgaver(Temagruppe.valueOf(temagruppe.toUpperCase()),
                                        CookieUtil.getSaksbehandlersValgteEnhet(httpRequest))
                    }
                }.map { mapOppgave(it) }
    }

    @GetMapping("/tildelt")
    fun finnTildelte() =
            tilgangkontroll
                    .check(Policies.tilgangTilModia)
                    .get(Audit.describe(READ, Henvendelse.Oppgave.Tildelte)) {
                        oppgaveBehandlingService.finnTildelteOppgaverIGsak()
                                .map { mapOppgave(it) }
                    }

    @GetMapping("/oppgavedata/{oppgaveId}")
    fun getOppgaveData(@PathVariable("oppgaveId") oppgaveId: String): OppgaveDTO =
            tilgangkontroll
                    .check(Policies.tilgangTilModia)
                    .get(Audit.describe(READ, Henvendelse.Oppgave.Metadata, AuditIdentifier.OPPGAVE_ID to oppgaveId)) {
                        mapOppgave(oppgaveBehandlingService.hentOppgave(oppgaveId))
                    }

    private fun lagLeggTilbakeRequest(request: LeggTilbakeRequest, valgtEnhet: String): LeggTilbakeOppgaveIGsakRequest? {
        require(request.oppgaveId != null)
        val baseRequest = LeggTilbakeOppgaveIGsakRequest()
                .withOppgaveId(request.oppgaveId)
                .withSaksbehandlersValgteEnhet(valgtEnhet)
        return when (request.type) {
            LeggTilbakeAarsak.Innhabil -> baseRequest
                    .withBeskrivelse(AARSAK_PREFIX + "inhabil")
            LeggTilbakeAarsak.FeilTema -> {
                require(request.temagruppe != null)
                return baseRequest
                        .withBeskrivelse(AARSAK_PREFIX + "feil temagruppe")
                        .withTemagruppe(request.temagruppe)
            }
            LeggTilbakeAarsak.AnnenAarsak -> {
                require(request.beskrivelse != null)
                return baseRequest
                        .withBeskrivelse(AARSAK_PREFIX + request.beskrivelse)
            }
        }
    }
}

data class OppgaveDTO(
        val oppgaveId: String,
        val traadId: String?,
        val fødselsnummer: String?,
        val erSTOOppgave: Boolean
)
private fun mapOppgave(oppgave: Oppgave) = OppgaveDTO(
        oppgave.oppgaveId,
        oppgave.henvendelseId,
        oppgave.fnr,
        oppgave.erSTOOppgave
)

private fun handterRuntimeFeil(exception: RuntimeException): RuntimeException {
    logger.error("Feil ved legging av oppgave tilbake til GSAK", exception)
    return exception
}

enum class LeggTilbakeAarsak {
    Innhabil,
    FeilTema,
    AnnenAarsak
}

data class LeggTilbakeRequest(
        val type: LeggTilbakeAarsak,
        val oppgaveId: String,
        val temagruppe: Temagruppe?,
        val beskrivelse: String?,
        val traadId: String?
)
